streaming-payments.md

```diff
@@ -0,0 +1,375 @@
+const fetch = require('node-fetch')
+const uuid = require('uuid/v4')
+function base64 (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+function hash (secret) { return crypto.createHash('sha256').update(secret).digest() }
+function hmac (secret, input) { return crypto.createHmac('sha256', secret).update(input).digest() }
+
+const plugin = new Plugin({
+  secret: 'sndb5JDdyWiHZia9zv44zSr2itRy1',
+  account: 'rGtqDAJNTDMLaNNfq1RVYgPT8onFMj19Aj',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+let counter = 0
+
+function sendTransfer (obj) {
+  obj.id = uuid()
+  obj.from = plugin.getAccount()
+  // to
+  obj.ledger = plugin.getInfo().prefix
+  // amount
+  // executionCondition
+  obj.expiresAt = new Date(new Date().getTime() + 1000000).toISOString()
+  return plugin.sendTransfer(obj).then(function () {
+    return obj.id
+  })
+}
+
+plugin.connect().then(function () {
+  return fetch('http://localhost:8000/')
+}).then(function (inRes) {
+  inRes.body.pipe(process.stdout)
+  const payHeaderParts = inRes.headers.get('Pay').split(' ')
+  console.log(payHeaderParts)
+  // e.g. Pay: 1 test.crypto.xrp.asdfaqefq3f.26wrgevaew SkTcFTZCBKgP6A6QOUVcwWCCgYIP4rJPHlIzreavHdU
+  setInterval(function () {
+    const ilpPacketContents = {
+      account: payHeaderParts[1] + '.' + (++counter),
+      amount: '1',
+      data: ''
+    }
+    const fulfillment = hmac(Buffer.from(payHeaderParts[2], 'base64'), ilpPacketContents.account)
+    const condition = hash(fulfillment)
+    sendTransfer({
+      to: payHeaderParts[1],
+      amount: '1',
+      executionCondition: base64(condition),
+      ilp: base64(IlpPacket.serializeIlpPayment(ilpPacketContents))
+    }).then(function () {
+      // console.log('transfer sent')
+    }, function (err) {
+      console.error(err.message)
+    })
+  }, 500)
+})
+```
+
+As you can see, it parses the `Pay` header, and then sends one XRP per 500ms. Try it out!
+
+```sh
+$ node ./tool.js
+```
+
+After some startup time, you should see letters being printed.
+
+## Using a trustline
+
+Getting one letter per 500ms is not very fast. It would be nice if we could stream the money faster, so that
+the content arrives faster! For this, we can add a ledger to the shop. The tool opens an account on this ledger,
+and then pays for letters from its account at the shop's ledger, which will be much faster
+than paying via the XRP ledger.
+
+The shop's ledger will expose the Bilateral Transfer Protocol (BTP), which is an optimization of the Ledger Plugin Interface (LPI)
+that we already saw in the Letter Shop tutorial. It is symmetrical over a WebSocket, so the client can send the same messages as the
+server, and this will have the same effect, but in the opposite direction. For instance, both parties can send a transfer to the
+other party in the form of a 'PREPARE' packet, and both parties can also fulfill transfers from the other party by sending a
```

Comment:
This is confusing to explain here. If I'm a sending client, why would I be sending the same messages as the server?

Reply:
Right, I should spend more words on that, probably. An ILP sender can be a BTP server, and a n ILP receiver can be a BTP client. So apart from (ILP-)sending (BTP-)clients, we can have receiving clients, sending servers, and receiving servers.